### DISCUSSION TITLE
Issues while Running OpenVSLAM on Ros2 with docker

### DISCUSSION THREAD

**Discussion by Udaykiran87** [stella-cv/stella_vslam#122] [Q&A] [answered]
Hello Community,

I was trying to run OpenVSLAM on ros2 with docker.
I followed two methods and in both i got errors. I do not know the correct steps to follow, below steps were based on my understandings.

METHOD 1:
I was going through following url for running on ros2.
https://openvslam-community.readthedocs.io/en/latest/ros2_package.html#installation

but this did not clearly mentioned how to run it on docker.

However I tried below steps(which i think is wrong) and came across error mentioned in the last step.

s1. git clone below:
https://github.com/OpenVSLAM-Community/openvslam_ros

s2:
cd openvslam_ros
docker build -t openvslam-ros-f .

s3:
xhost +local:
docker run -it --rm -e DISPLAY=$DISPLAY --runtime=nvidia -v /tmp/.X11-unix/:/tmp/.X11-unix:ro openvslam-ros-f

s4:
mkdir -p ~/ros2_ws/src
cd ~/ros2_ws/src
git clone -b ${ROS_DISTRO} --single-branch https://github.com/ros-perception/image_common.git

I got below message:
root@5bac2ccbf130:~/ros2_ws/src# git clone -b ${ROS_DISTRO} --single-branch https://github.com/ros-perception/image_common.git
Cloning into 'image_common'...
warning: Could not find remote branch foxy to clone.
fatal: Remote branch foxy not found in upstream origin

So i tried with
git clone -b ros2 --single-branch https://github.com/ros-perception/image_common.git
git clone -b ros2 --single-branch https://github.com/ros-perception/vision_opencv.git

s5:
cd ~/ros2_ws
git clone https://github.com/ros2/demos.git
cp -r demos/image_tools src/
rm -rf demos

s6:
cd cd ~/ros2_ws
colcon build --symlink-install --cmake-args -DUSE_PANGOLIN_VIEWER=ON -DUSE_SOCKET_PUBLISHER=OFF

process stopped with below error message:
root@5bac2ccbf130:~/ros2_ws# colcon build --symlink-install --cmake-args -DUSE_PANGOLIN_VIEWER=ON -DUSE_SOCKET_PUBLISHER=OFF
Starting >>> cv_bridge
Starting >>> camera_calibration_parsers
Starting >>> image_transport
Starting >>> image_geometry
--- stderr: cv_bridge                                                                                                                                             
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR python3)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:32 (find_package)

---
Failed   <<< cv_bridge [2.65s, exited with code 1]
Aborted  <<< image_geometry [4.57s]                                                                                                                 
Aborted  <<< camera_calibration_parsers [5.39s]                                                                       
Aborted  <<< image_transport [5.51s]                            

Summary: 0 packages finished [6.16s]
  1 package failed: cv_bridge
  3 packages aborted: camera_calibration_parsers image_geometry image_transport
  4 packages had stderr output: camera_calibration_parsers cv_bridge image_geometry image_transport
  6 packages not processed

Could you please tell me the correct steps to run openvslam on ros2 with docker.

METHOD 2:
I tried to follow below thread which i thought would be useful, but i got errors:
https://github.com/xdspacelab/openvslam/issues/432

Steps followed are as below:

s1:
# Clone the repository
cd $HOME
git clone -b ros2-base --single-branch https://github.com/mirellameelo/openvslam.git

s2: (I updated dockerfile first line as FROM ros:foxy instead of FROM osrf/ros:dashing-desktop and commented the last line of the docker file)
# to build
cd path/to/dockerfile
docker build -t openvslam-desktop -f Dockerfile.desktop .
# after built, run
docker run -it --rm -e DISPLAY=$DISPLAY --runtime=nvidia -v /tmp/.X11-unix/:/tmp/.X11-unix:ro openvslam-desktop

s3:
/bin/bash -c "source /ros_entrypoint.sh && cd $HOME/openvslam/ros2 && colcon build --symlink-install"

but got below error:
Starting >>> cv_bridge
Starting >>> image_transport
Starting >>> camera_calibration_parsers
Starting >>> image_geometry
--- stderr: cv_bridge                                                                                                                                            
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR python3)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:32 (find_package)

---
Failed   <<< cv_bridge [2.77s, exited with code 1]
Aborted  <<< image_geometry [4.59s]                                                                                                                
Aborted  <<< camera_calibration_parsers [5.51s]                                                                      
Aborted  <<< image_transport [5.69s]                           

Summary: 0 packages finished [6.32s]
  1 package failed: cv_bridge
  3 packages aborted: camera_calibration_parsers image_geometry image_transport
  1 package had stderr output: cv_bridge
  5 packages not processed

s3:
source /opt/ros/foxy/setup.bash
source $HOME/openvslam/ros2/install/setup.bash
/bin/bash -c "source /ros_entrypoint.sh && cd $HOME/openvslam/ros2 && colcon build --symlink-install"

but got bellow error again:
Starting >>> cv_bridge
Starting >>> image_transport
Starting >>> camera_calibration_parsers
Starting >>> image_geometry
--- stderr: cv_bridge                                                                                                                                                        
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR python3)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:32 (find_package)

---
Failed   <<< cv_bridge [2.35s, exited with code 1]
Aborted  <<< image_transport [32.2s]                                                                                                                               
Aborted  <<< image_geometry [34.7s]                                                                                             
Aborted  <<< camera_calibration_parsers [37.2s]                                  

Summary: 0 packages finished [37.8s]
  1 package failed: cv_bridge
  3 packages aborted: camera_calibration_parsers image_geometry image_transport
  2 packages had stderr output: cv_bridge image_transport
  5 packages not processed

Could you please help or guide me to resolve the issue.

Thanks and Regards,
Udaykiran Patnaik

**Comment by ymd-stella** [comment, marked as answer]
No steps after s4 are required. All necessary packages are already installed, please read the dockerfile.

**Reply by Udaykiran87** [reply]
Thanks for your reply. So instead of ${ROS_DISTRO}=foxy, I should use ros2 in s4? Because default ${ROS_DISTRO} gave error as below in s4 while downloading repositories of image_common and vision_opencv. Is my understanding correct?

```
Cloning into 'image_common'...
warning: Could not find remote branch foxy to clone.
fatal: Remote branch foxy not found in upstream origin
```

**Reply by ymd-stella** [reply]
Read openvslam_ros/Dockerfile. If it completes successfully, all the necessary packages are already installed and openvslam_ros is already built.

**Reply by Udaykiran87** [reply]
Thank you for your reply. Since docker build is successful in my case i assume that i do not need s4 to explicitly download image_common and vision_opencv. Please correct me if i am not wrong.

In case i need to use USB camera as a image source, i need to download a repository of demos and pick image_tools module as per [this link](https://openvslam-community.readthedocs.io/en/latest/ros2_package.html#installation). Which i did. Then once i download everything inside src folder i think i am supposed to build this package.

If i dont build and use below command  "**ros2 run image_tools cam2image**" then i get error as **Package 'image_tools' not found**
Is there some different package i need to run to use USB camera as source. When i run "**ros2 pkg list**" i did not find image_tools in the output.
So i assume i need to build newly downloaded image_tools package.
So when i try to build using "**colcon build --symlink-install --cmake-args -DUSE_PANGOLIN_VIEWER=ON -DUSE_SOCKET_PUBLISHER=OFF**"

but i get below **error**:

root@1d333d584b94:~/ros2_ws# colcon build --symlink-install --cmake-args -DUSE_PANGOLIN_VIEWER=ON -DUSE_SOCKET_PUBLISHER=OFF
```
Starting >>> camera_calibration_parsers
Starting >>> cv_bridge
Starting >>> image_geometry
Starting >>> image_transport
--- stderr: cv_bridge                                                                                                                                            
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR python3)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:32 (find_package)


---
Failed   <<< cv_bridge [2.84s, exited with code 1]
Aborted  <<< image_geometry [4.89s]                                                                                                                
Aborted  <<< camera_calibration_parsers [5.81s]                                                                      
Aborted  <<< image_transport [5.87s]                           

Summary: 0 packages finished [6.51s]
  1 package failed: cv_bridge
  3 packages aborted: camera_calibration_parsers image_geometry image_transport
  4 packages had stderr output: camera_calibration_parsers cv_bridge image_geometry image_transport
  5 packages not processed
```

Please help me to understand if my steps are correct or not.
Also, when i tried to run video example using "ros2 run publisher" command i got error as unable to find publisher package.

Thanks and Regards,
Udaykiran Patnaik.

**Reply by ymd-stella** [reply]
Updated the outdated section of openvslam_ros/Dockerfile. Please try again. If you still have questions, please ask.  Contributions for additional documentation are also welcome.

**Reply by ymd-stella** [reply]
> In case i need to use USB camera as a image source, i need to download a repository of demos and pick image_tools module as per this link. 

Can't install with apt?

**Reply by ymd-stella** [reply]
If you really want to build from source code, please install boost as shown in the error message.
ROS can automatically install the dependency package with the following command.

```
cd ~/ros2_ws
apt update
rosdep install -y -i --from-paths src
```

**Reply by Udaykiran87** [reply]
> Updated the outdated section of openvslam_ros/Dockerfile. Please try again. If you still have questions, please ask. Contributions for additional documentation are also welcome.

Sure I will try and let you know my results and will sure contribute.

**Reply by Udaykiran87** [reply]
> > In case i need to use USB camera as a image source, i need to download a repository of demos and pick image_tools module as per this link.
> 
> Can't install with apt?

I have not tried yet. Need to check.

**Reply by Udaykiran87** [reply]
I updated the docker file to include image_tools as below:
```
# ROS2
RUN set -x && \
  apt-get update -y -qq && \
  : "install ROS2 packages" && \
  apt-get install -y -qq \
    ros-${ROS_DISTRO}-image-transport \
    ros-${ROS_DISTRO}-cv-bridge \
    ros-${ROS_DISTRO}-image-tools \
    python3-pip \
    python3-colcon-common-extensions && \
  pip3 install -U \
    argcomplete && \
  : "remove cache" && \
  apt-get autoremove -y -qq && \
  rm -rf /var/lib/apt/lists/*
```

Then i ran below command:
**ros2 run image_tools cam2image**

But i got error as below:
root@4236498f2dd3:/ros2_ws# ros2 run image_tools cam2image
**terminate called after throwing an instance of 'class_loader::LibraryLoadException'
  what():  Could not load library dlopen error: libnvll.so: cannot open shared object file: No such file or directory, at /tmp/binarydeb/ros-galactic-rcutils-4.0.2/src/shared_library.c:99**

I wonder any other dependencies are missing or what?

Could you please help me

Thanks and Regards,
Udaykiran Patnaik

**Reply by ymd-stella** [reply]
I was not able to reproduce this in my environment. Insufficient information.

**Reply by Udaykiran87** [reply]
Hello @ymd-stella 

Thank you so much for you support. I really appreciate. 
It was my bad.
Since i am running Open VSLAM inside Jetson AGX, i have to specify my input device video parameter correctly.
My working command is as below:
**docker run -it --rm --device /dev/video0 -e DISPLAY=$DISPLAY -v /tmp/.X11-unix/:/tmp/.X11-unix:ro openvslam-ros-g**

However i ran into another problem.
I followed below steps:
s1:
ros2 run image_tools cam2image
s2:
ros2 run openvslam_ros run_slam -v ./orb_vocab.fbow -c ../tmp/openvslam/example/aist/config/camera.yaml

**Error Message:**
```
root@f9e0166818c3:/ros2_ws# ros2 run openvslam_ros run_slam -v ./orb_vocab.fbow -c ../tmp/openvslam/example/aist/config/camera.yaml
[2021-06-09 13:59:14.440] [I] config file loaded: ../tmp/openvslam/example/aist/config/camera.yaml
[2021-06-09 13:59:14.441] [I] 
  ___               __   _____ _      _   __  __ 
 / _ \ _ __  ___ _ _\ \ / / __| |    /_\ |  \/  |
| (_) | '_ \/ -_) ' \\ V /\__ \ |__ / _ \| |\/| |
 \___/| .__/\___|_||_|\_/ |___/____/_/ \_\_|  |_|
      |_|                                        

Copyright (C) 2019,
National Institute of Advanced Industrial Science and Technology (AIST)
All rights reserved.

This is free software,
and you are welcome to redistribute it under certain conditions.
See the LICENSE file.

Camera:
  name: logicool
  setup: monocular
  model: perspective
  fx: 820.68303564
  fy: 820.51434565
  cx: 310.87149613
  cy: 231.63407464
  k1: 0.037160015
  k2: 1.96485337
  p1: 0.00136894818
  p2: 0.0084237012
  k3: -16.9908964
  fps: 30.0
  cols: 640
  rows: 480
  color_order: RGB
Feature:
  max_num_keypoints: 1000
  ini_max_num_keypoints: 2000
  scale_factor: 1.2
  num_levels: 8
  ini_fast_threshold: 20
  min_fast_threshold: 7
Mapping:
  baseline_dist_thr_ratio: 0.02
PangolinViewer:
  keyframe_size: 0.07
  keyframe_line_width: 1
  graph_line_width: 1
  point_size: 2
  camera_size: 0.08
  camera_line_width: 3
  viewpoint_x: 0
  viewpoint_y: -0.65
  viewpoint_z: -1.9
  viewpoint_f: 400

[2021-06-09 13:59:14.441] [I] loading ORB vocabulary: ./orb_vocab.fbow
[2021-06-09 13:59:14.505] [I] startup SLAM system
[2021-06-09 13:59:14.506] [I] start mapping module
[2021-06-09 13:59:14.506] [I] start global optimization module
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X11 Error: BadMatch (invalid parameter attributes)
X11 Error: BadValue (integer parameter out of range for operation)
terminate called after throwing an instance of 'std::runtime_error'
  what():  Pangolin X11: Failed to create an OpenGL context
*** Aborted at 1623247154 (unix time) try "date -d @1623247154" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x4d) received by PID 77 (TID 0x7f91d0a720) from PID 77; stack trace: ***
    @       0x7fab405f10 (unknown)
    @       0x7fab5e36c0 ([vdso]+0x6bf)
    @       0x7faa39e138 gsignal
    @       0x7faa38ad68 abort
    @       0x7faa5a08ac __gnu_cxx::__verbose_terminate_handler()
    @       0x7faa59e1ec (unknown)
    @       0x7faa59e250 std::terminate()
    @       0x7faa59e544 __cxa_throw
    @       0x7fab4f2a30 pangolin::CreateGlContext()
    @       0x7fab4f2e3c pangolin::X11GlContext::X11GlContext()
    @       0x7fab4f32d4 pangolin::CreateX11WindowAndBind()
    @       0x7fab4f4828 pangolin::RegisterX11WindowFactory()::X11WindowFactory::Open()
    @       0x7fab4b6180 pangolin::CreateWindowAndBind()
    @       0x7fab56d11c pangolin_viewer::viewer::run()
    @       0x557028e094 _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZ8trackingRKSt10shared_ptrIN9openvslam6configEERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESG_bSG_bEUlvE_EEEEE6_M_runEv
    @       0x7faa5cbc2c (unknown)
    @       0x7faac904fc start_thread
    @       0x7faa43b67c (unknown)

**Note:
I am using the same vocabulary and .yaml file for my webcam which was using for non ros version last month. But only difference is that this time i am using pangolin instead  socket viewer.**

Kindly help me to understand the issue.

Thanks and Regards,
Udaykiran Patnaik.

```

**Reply by Udaykiran87** [reply]
After doing some search i found some related thread. As per [this](https://github.com/xdspacelab/openvslam/issues/432#issuecomment-688324514) thread, i used below command to run container.
s1.
xhost +local:
s2.
**docker run -it --rm --device /dev/video0 -e DISPLAY=$DISPLAY --runtime=nvidia --net=host -v /tmp/.X11-unix/:/tmp/.X11-unix:ro openvslam-ros-g**

but after this when i used "**ros2 run image_tools cam2image"** i got below error. Seems to be some compatibility issue with ros-galactic.
Note: I already have Nvidia Docker 2.

```
terminate called after throwing an instance of 'class_loader::LibraryLoadException'
  what():  Could not load library dlopen error: libnvll.so: cannot open shared object file: No such file or directory, at /tmp/binarydeb/ros-galactic-rcutils-4.0.2/src/shared_library.c:99

```

**Comment by Udaykiran87** [comment]
So, I am confused to use **--runtime=nvidia** or not in my docker run command in my jetson agx xavier (ubuntu 18.04).

Summary of error:
If i use **--runtime=nvidia** in docker run i can not run `"ros2 run image_tools cam2image"` ,
but if don't use  **--runtime=nvidia** then i can  run `"ros2 run image_tools cam2image" `but after that when i run 
`ros2 run openvslam_ros run_slam -v ./orb_vocab.fbow -c ../tmp/openvslam/example/aist/config/camera.yaml `
i get below error.

```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X11 Error: BadMatch (invalid parameter attributes)
X11 Error: BadValue (integer parameter out of range for operation)
terminate called after throwing an instance of 'std::runtime_error'
  what():  Pangolin X11: Failed to create an OpenGL context
```

Thanks and Regards,
Udaykiran Patnaik.

**Reply by ymd-stella** [reply]
The command I am using is as follows:

```
touch /tmp/.docker.xauth && xauth nlist $DISPLAY | sed -e 's/^..../ffff/' | xauth -f /tmp/.docker.xauth nmerge - && docker run -it --rm --name openvslam-ros2-test -e DISPLAY=$DISPLAY -e XDG_RUNTIME_DIR=${XDG_RUNTIME_DIR} -e XAUTHORITY=/tmp/.docker.xauth -v /tmp/.X11-unix/:/tmp/.X11-unix:ro -v ${XDG_RUNTIME_DIR}:${XDG_RUNTIME_DIR} -v /tmp/.docker.xauth:/tmp/.docker.xauth --net=host openvslam-ros2-test
```

**Reply by ymd-stella** [reply]
It is also recommended to use SocketViewer.

**Reply by Udaykiran87** [reply]
@ymd-stella : Thank you so much for your reply. Actually i misunderstood that with ROS2 only PANGOLIN_VIEWER possible.
I will try with -DUSE_PANGOLIN_VIEWER=OFF and -DUSE_SOCKET_PUBLISHER=ON.

I hope above change is sufficient in the existing docker file inside ros repo to use socketviewer. Please let me know if any additional changes are required for socket viewer for ros2.

Last month with socketviewer i was successfully able to run OpenVSLAM with docker but without Ros2 in my jetson xavier. I will let you my result.

Thanks and Regards,
Udaykiran patnaik.

**Reply by Udaykiran87** [reply]
> The command I am using is as follows:
> 
> ```
> touch /tmp/.docker.xauth && xauth nlist $DISPLAY | sed -e 's/^..../ffff/' | xauth -f /tmp/.docker.xauth nmerge - && docker run -it --rm --name openvslam-ros2-test -e DISPLAY=$DISPLAY -e XDG_RUNTIME_DIR=${XDG_RUNTIME_DIR} -e XAUTHORITY=/tmp/.docker.xauth -v /tmp/.X11-unix/:/tmp/.X11-unix:ro -v ${XDG_RUNTIME_DIR}:${XDG_RUNTIME_DIR} -v /tmp/.docker.xauth:/tmp/.docker.xauth --net=host openvslam-ros2-test
> ```

Thank you for sharing.

**Reply by ymd-stella** [reply]
> I hope above change is sufficient in the existing docker file inside ros repo to use socketviewer. Please let me know if any additional changes are required for socket viewer for ros2.

I've added Dockerfile.socket, please use it. Document is [here](https://openvslam-community.readthedocs.io/en/latest/docker.html#docker-image-of-server).

**Comment by Udaykiran87** [comment]
@ymd-stella 
Thank you so much. I was successfully able to run OpenVSLAM+Ros2+Docker.scocket. I tried with with aist_living_lab_1 for mapping and aist_living_lab_2 for localizing as well as USB camera. I had to modify openvslam_video publisher to modify the   rclcpp::WallRate pub_rate(video.get(cv::CAP_PROP_FPS)); to   rclcpp::WallRate pub_rate(2.5); otherwise frames were lost.

While with above videos, mapping and localization looked ok with naked eyes but i was getting below error messages for both mapping  and localizing on my screen. Is this  below behavior is accepted or some issues.

**Snippet of Mapping Error message:**
```
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400010.648465751] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400010.814029537] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400011.209060863] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400011.630941740] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400012.211281383] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400012.912941009] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400013.315274708] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400013.856253101] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623400013.973217177] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
```
**Snippet for localization error message:**
```
       at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403888.530545885] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403889.892741121] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403890.753891127] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403891.231569402] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403891.835575619] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp
[ERROR] [1623403891.967933801] [run_slam]: Transform failed: "camera_link" passed to lookupTransform argument target_frame does not exist. 
Warning: Invalid frame ID "camera_link" passed to canTransform argument target_frame - frame does not exist
         at line 156 in /tmp/binarydeb/ros-galactic-tf2-0.17.2/src/buffer_core.cpp

```

Thanks and Regards,
Udaykiran Patnaik.

**Reply by ymd-stella** [reply]
If tf is not needed, set the parameter publish_tf to false. Otherwise, you will need to provide a tf from odom to camera_link. Usually, the tf from odom to base_link is calculated by wheel odometry, and the tf from base_link to camera_link is specified by static_transform_publisher.

**Reply by Udaykiran87** [reply]
@ymd-stella 
Ok, Thank you so much. I will update and check.

Thanks and Regards,
Udaykiran Patnaik.

**Reply by Udaykiran87** [reply]
@ymd-stella 
I updated the code as suggested by you and the error is gone.
